src/ocean/application/Daemon.d

```diff
@@ -0,0 +1,769 @@
+          built-in commands:
+            * show_version
+            * show_build_info
+            * reopen_files: reopens the specified files (must be registered with
+              the open files registry; see above)
+        * Creation of a PID lock file, to prevent multiple instances of the
+          application from starting in the same directory.
+        * Epoll and task scheduler setup.
+        * A main Task instance that the application's main logic is run in.
+
+    Usage example:
+        See Daemon class' documented unittest
+
+    Copyright:
+        Copyright (c) 2018 dunnhumby Germany GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.application.Daemon;
+
+import ocean.transition;
+
+/// ditto
+public abstract class Daemon
+{
+    import ocean.core.Enforce;
+    import ocean.core.ExceptionDefinitions : IOException;
+    import ocean.core.Verify;
+    import ocean.io.device.File;
+    import ocean.io.Stdout;
+    import ocean.task.Scheduler;
+    import ocean.task.Task;
+    import ocean.text.Arguments;
+    import ocean.util.app.ExitException;
+    import ocean.util.config.ConfigParser;
+    import ConfigFiller = ocean.util.config.ConfigFiller;
+    import ocean.util.log.Appender;
+    import ocean.util.log.AppendFile;
+    import LogConfig = ocean.util.log.Config;
+    import ocean.util.log.LayoutDate;
+    import ocean.util.log.Logger;
+    import ocean.util.log.Stats;
+    import ocean.sys.Stats;
+
+    import ConfigOverrides = ocean.application.components.ConfigOverrides;
+    import ocean.application.components.OpenFiles;
+    import ocean.application.components.PidLock;
+    import ocean.application.components.Signals;
+    import TaskScheduler = ocean.application.components.TaskScheduler;
+    import ocean.application.components.Timers;
+    import ocean.application.components.UnixSocketCommands;
+    import Version = ocean.application.components.Version;
+
+    /// Settings passed to the constructor.
+    public struct Settings
+    {
+        /// Application name. Mandatory.
+        istring name;
+
+        /// Application description string.
+        istring desc;
+
+        /// How the program is supposed to be invoked.
+        istring usage;
+
+        /// Long description of what the program does and how to use it.
+        istring help;
+
+        /// Application version info. Mandatory.
+        Version.VersionInfo ver;
+
+        /// Configuration files to parse.
+        istring[] config_files = [ "etc/config.ini" ];
+
+        /// If true, errors when parsing configuration files are treated as
+        /// fatal -- the app will not start.
+        bool config_errors_fatal = true;
```

Comment:
Does anyone have a real use case for setting this to `false`? (This would be the equivalent to the old "loose config parsing".)

Reply:
Some apps, most notably `thruster` & `propulsor`, have the ability to reload their configurations on receipt of a signal (or a command via the UNIX socket). I don't know what would happen if an error was made in the config file and then the reload command was sent. But supporting "loose config parsing" to handle this situation is anyway not a proper solution. So I don't have any strong arguments for keeping this boolean variable.

Reply:
Ah right, good point. I'd be interested to hear how that works in your apps, as this is something that I was wondering about adding (at some point) as a general feature of the daemon base class.

Reply:
Looking at the config re-parsing code from one of these apps, it does the following on receipt of SIGHUP:
```D
this.config_ext.preRun(this, null);
```
This triggers the reload of all config files, and calls the app-defined behaviour in `processConfig` again.

Reply:
This would be a nice behaviour to support "out of the box".

Reply:
Ok so I've added support for config reloading. A suggestion for `config_errors_fatal`: we say that this only applies at startup. When the config file is parsed in response to the unix socket command, parser errors are logged, but do not cause the application to exit.

Reply:
@gavin-norman-sociomantic : The app I maintained does. It was desirable with split config files, since you could want to test a new version which removes a config parameter used everywhere. Think of:

```
[LOG]
mode = append
file_size = 100_000
```

In your common config file, and both field being required. If you want to test, say, a new version of Ocean which removes `file_size`, you either have to specify the whole configuration in its own file and use that or rely on loose parsing. I choose to do the later (you can look at the commit / comment in the app in question).

However if someone relies on this, @mihails-strasuns ' solution means that they essentially can't reload their config. Doesn't sound too appealing.